frontend/catalyst/ag_primitives.py

```diff
@@ -411,6 +419,73 @@ def while_stmt(loop_test, loop_body, get_state, set_state, nonlocals, symbol_nam
     set_state(results)
 
 
+def _logical_op(*args, jax_fn, python_fn):
+    fallback = False
+
+    values = [f() for f in args]
+
+    def _non_scalar_tracer(x) -> bool:
+        return isinstance(x, DynamicJaxprTracer) and reduce(lambda a, b: a * b, x.shape, 1) != 1
+
+    if any(map(_non_scalar_tracer, values)):
+        raise AutoGraphError(
+            "Using logical operations (and/or/not) with non-scalar tracers is not supported\n"
+            "Consider using `jax.numpy.logical_and/..logical_or/..logical_not` functions.\n"
+            f"arguments provided were: {args}"
+        )
```

Comment:
+1 I would also like to know this. By allowing the `jnp.array(True) and jnp.array([False])` syntax we already go beyond what the original Python allows. So how far should we go here?

Reply:
>By allowing the jnp.array(True) and jnp.array([False]) syntax we already go beyond what the original Python allows.

What do you mean? I think that example is supported with Python `and`

Reply:
Hey all, I just spoke to @glassnotes and the preference here is to preserve the same behaviour as JAX and NumPy -- that is, the following error should be raised with AutoGraph, and the user would need to use `jnp.all` explicitly:

```python
>>> jnp.array([0, 1]) and jnp.array([1, 2])
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
>>> jnp.all(jnp.array([0, 1]), jnp.array([1, 2]))
False
```

Reply:
This is actually more complicated than it appears. @grwlf we're still trying to understand what the best approach might be.

Reply:
In addition to https://github.com/PennyLaneAI/catalyst/pull/325#discussion_r1373500085, there is no need to prevent the conversion from happening when the arrays are non-scalar.

> \>>> bool(jnp.array([0, 1]) and jnp.array([1, 2]))
> ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

We realized the error message here is not very helpful, because it comes from the `__bool__` function being called on the LHS operand, not from ANDing together two non-scalar arrays. `jnp.all(jnp.array([0, 1]) and jnp.array([1, 2]))` for example does not work!

Reply:
My original concern was related to the degree we want to emulate the pure-Python behavior. Without the check we have

``` python
def f(a, b):
  return a and b

>>> qjit(autograph=True)(f)(jnp.array([0,1]), jnp.array([1,1]))
array([False,  True])

>>> f(jnp.array([0,1]), jnp.array([1,1]))
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

```

Reply:
fyi: The warning is removed in the current version @dime10 please check